File: backend/app/api/v1/referrals.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from decimal import Decimal

from app.core.database import get_db
from app.core.deps import RoleChecker
from app.models.models import User, AmbassadorProfile, Referral, Order
from app.schemas.schemas import AmbassadorStats, AmbassadorProfileResponse

router = APIRouter()


@router.get("/profile", response_model=AmbassadorProfileResponse)
def get_ambassador_profile(
    db: Session = Depends(get_db),
    current_user: User = Depends(RoleChecker(["ambassador"]))
):
    """Retrieve current ambassador profile settings and code details."""
    profile = db.query(AmbassadorProfile).filter(AmbassadorProfile.user_id == current_user.id).first()
    if not profile:
        raise HTTPException(status_code=404, detail="Ambassador profile not initialized")
    return profile


@router.get("/stats", response_model=AmbassadorStats)
def get_ambassador_stats(
    db: Session = Depends(get_db),
    current_user: User = Depends(RoleChecker(["ambassador"]))
):
    """Compute performance dashboard aggregates: total clicks, referrals, conversions, and balances."""
    profile = db.query(AmbassadorProfile).filter(AmbassadorProfile.user_id == current_user.id).first()
    if not profile:
        raise HTTPException(status_code=404, detail="Ambassador profile not found")

    referrals = db.query(Referral).filter(Referral.ambassador_id == current_user.id).all()
    
    # Calculate stats
    registrations = len(referrals)
    converted = sum(1 for r in referrals if r.status in ["order_placed", "paid"])
    
    # Let's mock click rates (say 4 clicks per registration for realistic dashboard representation)
    simulated_clicks = registrations * 4 + 12

    total_earnings = sum(r.commission_earned for r in referrals)

    return AmbassadorStats(
        clicks=simulated_clicks,
        registrations=registrations,
        converted_orders=converted,
        total_earnings=Decimal(str(total_earnings)),
        current_balance=profile.balance,
        referrals=referrals
    )

File: backend/src/routes/referrals.ts
import { Router } from "express";
import { AmbassadorProfile, Referral, User } from "../models/models.js";
import { authenticateJWT, requireRoles } from "../middleware/auth.js";

const router = Router();

router.get("/profile", authenticateJWT, requireRoles(["ambassador"]), async (req, res) => {
  try {
    const profile = await AmbassadorProfile.findOne({ where: { user_id: req.user!.id } });
    if (!profile) {
      return res.status(404).json({ detail: "Ambassador profile not initialized" });
    }
    return res.json(profile);
  } catch (err: any) {
    console.error(err);
    return res.status(500).json({ detail: err.message || "Failed to load ambassador profile" });
  }
});

router.get("/stats", authenticateJWT, requireRoles(["ambassador"]), async (req, res) => {
  try {
    const profile = await AmbassadorProfile.findOne({ where: { user_id: req.user!.id } });
    if (!profile) {
      return res.status(404).json({ detail: "Ambassador profile not found" });
    }

    const referrals = await Referral.findAll({
      where: { ambassador_id: req.user!.id },
      include: [
        { model: User, as: "referred_user", attributes: { exclude: ["hashed_password"] } }
      ]
    });

    const registrations = referrals.length;
    const converted = referrals.filter(r => r.status === "order_placed" || r.status === "paid").length;
    const simulatedClicks = registrations * 4 + 12;

    const totalEarnings = referrals.reduce((sum, r) => sum + Number(r.commission_earned), 0);

    return res.json({
      clicks: simulatedClicks,
      registrations,
      converted_orders: converted,
      total_earnings: Number(totalEarnings.toFixed(2)),
      current_balance: Number(Number(profile.balance).toFixed(2)),
      referrals
    });
  } catch (err: any) {
    console.error(err);
    return res.status(500).json({ detail: err.message || "Failed to compute statistics" });
  }
});

export default router;
